### PR TITLE
Bug with claim types for groups in SPWebAppPolicy

### PR DESCRIPTION
**Pull request by BrianFarnhill**
The same fix that was made for SPServiceAppSecurity in #306 needs to be made to SPWebAppPolicy as well. Here is the fix.

- [X] Change details added to Unreleased section of readme.md?
- [X] Added/updated documentation and descriptions in .schema.mof files where appropriate?
- [X] Examples updated for both the single server and small farm templates in the examples folder?
- [X] New/changed code adheres to [Style Guidelines]?(https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md)?
- [X] [Unit and Integration tests](https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md) created/updated where possible?

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/sharepointdsc/319)

### CODE REVIEW

**Comment by ykuijs** [comment]
The Reviewable option is missing, reviewing using the default GitHub functionality

File: Modules/SharePointDSC/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1 [hunk 1]
```diff
@@ -342,7 +368,16 @@ function Set-TargetResource
...
+                    $userToDrop = $user.Username
+                    if ($wa.UseClaimsAuthentication -eq $true) {
+                        $isUser = Test-SPDSCIsADUser -IdentityName $user.Username
+                        if ($isUser -eq $true) {
+                            $userToDrop = (New-SPClaimsPrincipal -Identity $user.Username -IdentityType WindowsSamAccountName).ToEncodedString()
+                        } else {
+                            $userToDrop = (New-SPClaimsPrincipal -Identity $user.Username -IdentityType WindowsSecurityGroupName).ToEncodedString()
+                        }    
+                    }
+                    Remove-WebAppPolicy $wa.Policies userToDrop
```
Comment (ykuijs) [inline Modules/SharePointDSC/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1]: userToDrop must be $userToDrop

**Comment by ykuijs** [comment]
The scenario's that this resource currently supports are adding domain\user **and** claims formats. Both scenario's are very important to support, since even though the web app is configured as claims, when for example using Powershell you are still connecting using classic credentials. This was one of the reasons I did the refactor of the resource several weeks ago.

With this update, the domain\user scenario is removed. Secondly, I think this resource doesn't need updating. It already takes claims/non-claims and users/groups into account. When you add a group in claims format to the config, it will add the group in claims format correctly.

So therefore I would like to fully drop this PR and not merge it into the repo

**Comment by BrianFarnhill** [comment]
@ykuijs Not quiet mate. So yes it will let you do both right now, but the implementation for groups is 100% unusable. Claims for groups are formatted with the SID from AD, not the name, which means in the UI all you would see is the claim - same for the configuration itself as well, which makes managing it a nightmare as well - so we need something that gives us an option to supply a principal name and have it convert it to claims to be implemented. I've got a customer with exactly this scenario right now, which is what led me to this.

So I take your point about needing the non-claims versions though, so we need a work around for this. I figure we can do one of a couple of things:

1. We can add the claims and non-claims versions of accounts at the same time. This might wreak havoc with the test method though if only one type is there and not both - but logically speaking specifying that you desire a permission to be set for an account and having it set both the claim and non-claim versions for that account I think could make sense. 
2. We need to add an option that lets us specify if we want a user added as a claims user or a normal user so that we can throw in normal user names to keep it usable, while having it resolve the claims in the back end

We definitely need to do something with this though for the groups claims scenario - so we'll leave this PR open and work it here.

**Comment by ykuijs** [comment]
Personally I do not like option 1. It feels like overkill.

Option 2 is much better. I think the method that makes the most sense is to implement this, is to extend the current child class MSFT_SPWebPolicyPermissions to include an TokenFormat (possible values Claims and Classic) parameter. This parameter specifies if the Username has to be configured in Classic or Claims format. That way in the config you always specify the classic format and the resource converts it to the correct format.

What do you think about this option?

**Comment by BrianFarnhill** [comment]
@ykuijs Happy with that option, I'll re-code and ping you here when it's ready for review

**Comment by BrianFarnhill** [comment]
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/powershell/sharepointdsc/319#-KJsEof8-ldul-J9XA9L:-KJsEof8-ldul-J9XA9M:-608658491) (no related file):*
@ykuijs This one got a little out of hand when I started refactoring, but I think I'm pretty happy with where it landed. There are also a lot of changes to do with the PowerShell team style guidelines that we need to bring the module in line with (I'll be making changes like this as we go until I get time to do a run over the entire module for this) - ping me if you want to discuss any of it though. Good luck with the review!

---

*[Modules/SharePointDSC/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1, line 380 \[r1\]](https://reviewable.io:443/reviews/powershell/sharepointdsc/319#-KJkFijlZOcbbdtQG-Jb-r1-380:-KJsFGdg88LKv9K45z2B:-499146009) ([raw file](https://github.com/powershell/sharepointdsc/blob/1468de7199700deb84dcc675ede82ebffc55a806/Modules/SharePointDSC/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1#L380)):*
<details><summary><i>Previously, ykuijs (Yorick Kuijs) wrote…</i></summary>
> userToDrop must be $userToDrop
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/sharepointdsc/319)*

**Comment by ykuijs** [comment]
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/powershell/sharepointdsc/319#-KJsEof8-ldul-J9XA9L:-KK9xf8psv8NJ2uupwwT:-815310683) (no related file):*
<details><summary><i>Previously, BrianFarnhill (Brian Farnhill) wrote…</i></summary>
> @ykuijs This one got a little out of hand when I started refactoring, but I think I'm pretty happy with where it landed. There are also a lot of changes to do with the PowerShell team style guidelines that we need to bring the module in line with (I'll be making changes like this as we go until I get time to do a run over the entire module for this) - ping me if you want to discuss any of it though. Good luck with the review!
</details>
If you call this little ;-) One small comment/question. The rest looks excellent!

---

*[Modules/SharePointDsc/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1, line 442 \[r2\]](https://reviewable.io:443/reviews/powershell/sharepointdsc/319#-KK9vuQ8XhwxjeglDWmG:-KK9vuQ8XhwxjeglDWmH:120153787) ([raw file](https://github.com/powershell/sharepointdsc/blob/273366cabee113048e6a42212c90ed9b9345e387/Modules/SharePointDsc/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1#L442)):*
> ```PowerShell
>                     $polbinddiff = Compare-Object -ReferenceObject $policy.PolicyRoleBindings.Name -DifferenceObject $user.PermissionLevel
>                     if ($null -ne $polbinddiff) 
>                     {
> ```

You removed the RemoveAll line. I put this in to make sure you have a clean start. Without it, it is possible that if a user already has e.g. FullRead permissions, the FullControl permissions are added.....meaning you still aren't in the correct state.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/sharepointdsc/319)*

**Comment by BrianFarnhill** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[Modules/SharePointDsc/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1, line 442 \[r2\]](https://reviewable.io:443/reviews/powershell/sharepointdsc/319#-KK9vuQ8XhwxjeglDWmG:-KKB6RmpfhZfomSezll1:-1171940005) ([raw file](https://github.com/powershell/sharepointdsc/blob/273366cabee113048e6a42212c90ed9b9345e387/Modules/SharePointDsc/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1#L442)):*
<details><summary><i>Previously, ykuijs (Yorick Kuijs) wrote…</i></summary>
> You removed the RemoveAll line. I put this in to make sure you have a clean start. Without it, it is possible that if a user already has e.g. FullRead permissions, the FullControl permissions are added.....meaning you still aren't in the correct state.
</details>
Done. Good Pickup!

---

*[Modules/SharePointDSC/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1, line 380 \[r1\]](https://reviewable.io:443/reviews/powershell/sharepointdsc/319#-KJkFijlZOcbbdtQG-Jb-r1-380:-KKB6WpPKJ8qrD8SRNUW:-499146009) ([raw file](https://github.com/powershell/sharepointdsc/blob/1468de7199700deb84dcc675ede82ebffc55a806/Modules/SharePointDSC/DSCResources/MSFT_SPWebAppPolicy/MSFT_SPWebAppPolicy.psm1#L380)):*
<details><summary><i>Previously, BrianFarnhill (Brian Farnhill) wrote…</i></summary>
> Done.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/sharepointdsc/319)*

**Comment by BrianFarnhill** [comment]
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/powershell/sharepointdsc/319#-KJsEof8-ldul-J9XA9L:-KKBGvwM-UP7SKnFn2cE:-151074601) (no related file):*
<details><summary><i>Previously, ykuijs (Yorick Kuijs) wrote…</i></summary>
> If you call this little ;-) One small comment/question. The rest looks excellent!
</details>
Ready for another review mate

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/sharepointdsc/319)*

**Comment by ykuijs** [comment]
Reviewed 5 of 7 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/sharepointdsc/319)*

**Pull request merged**